File: verifier/python/test_vectors.py
"""
test_vectors.py — conformance test runner for the HumanAccepted receipt format.

Asserts that the Python reference verifier passes every vector in
vectors/v1/. The test runner is part of the spec so a translator
verifier (Go, Rust, etc.) can use the same vectors to assert its
own conformance.

Usage:
    python3 verifier/python/test_vectors.py
    # or:
    from verifier.python.test_vectors import run_all_vectors
    passed, failed = run_all_vectors()
    assert failed == 0, f"{failed} vectors failed"

Exit code: 0 if all vectors pass, 1 otherwise. Each failed vector
prints the verdict and reason to stderr.
"""

import json
import sys
from pathlib import Path

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

# The reference verifier is in the same directory. We import via
# package-relative path so the test runner works both as `python3
# test_vectors.py` (from this dir) and as `python3 -m
# verifier.python.test_vectors` (from the repo root).
HERE = Path(__file__).resolve().parent
if str(HERE) not in sys.path:
    sys.path.insert(0, str(HERE))

from verify import verify, _public_key_from_hex_or_pem  # type: ignore  # noqa: E402


# Vectors live at <repo>/vectors/v1/*.json. Resolve from this file.
REPO_ROOT = HERE.parent.parent
VECTORS_DIR = REPO_ROOT / "vectors" / "v1"
KEYS_FILE = VECTORS_DIR / "keys.json"


def _load_keys() -> dict:
    if not KEYS_FILE.exists():
        return {}
    return json.loads(KEYS_FILE.read_text(encoding="utf-8"))


def _iter_vectors():
    for path in sorted(VECTORS_DIR.glob("*.json")):
        if path.name == "keys.json":
            continue
        yield path


def _check_one(vector_path: Path, keys: dict) -> tuple[bool, str]:
    receipt = json.loads(vector_path.read_text(encoding="utf-8"))
    name = receipt.get("name", vector_path.name)

    # Structural pre-checks: version, signature shape, etc.
    if receipt.get("version") != 1:
        return False, f"{name}: expected version=1, got {receipt.get('version')!r}"
    sigs = receipt.get("signatures") or {}
    sig = sigs.get("tenant_ed25519")
    if not sig or not sig.startswith("ed25519:"):
        return False, f"{name}: signatures.tenant_ed25519 is not an ed25519:hex value"
    if "cf_attestation" in sigs and sigs["cf_attestation"] is not None:
        return False, f"{name}: cf_attestation must be null in v1"

    # The expected public key for this vector is in keys.json (we sign
    # with a deterministic per-vector key). If keys.json doesn't have an
    # entry, this vector is unsigned (e.g. a structural-only test).
    key_entry = keys.get(name)
    if not key_entry:
        # The vector has a placeholder sig. We only assert that the
        # verifier rejects it.
        # Use a dummy key to drive the verifier (the signature check
        # will fail because the placeholder isn't a real signature).
        dummy_key = _public_key_from_hex_or_pem("00" * 32)
        ok, reason = verify(receipt, dummy_key)
        if ok is True:
            return False, f"{name}: expected structural-only rejection, got valid"
        return True, f"rejected as expected (no keys.json entry): {reason}"

    # Full signature verification.
    pub = _public_key_from_hex_or_pem("ed25519:" + key_entry["public_key_hex"])
    ok, reason = verify(receipt, pub)
    if ok is not True:
        return False, f"{name}: verify() returned ok=False, reason={reason!r}"
    return True, reason


def run_all_vectors(verbose: bool = True) -> tuple[int, int]:
    """Run every vector in vectors/v1/. Returns (passed, failed)."""
    keys = _load_keys()
    passed = 0
    failed = 0
    for path in _iter_vectors():
        ok, detail = _check_one(path, keys)
        if ok:
            passed += 1
            if verbose:
                print(f"  ✓ {path.name}: {detail}")
        else:
            failed += 1
            if verbose:
                print(f"  ✗ {path.name}: {detail}", file=sys.stderr)
            else:
                print(f"  ✗ {path.name}", file=sys.stderr)
    return passed, failed


def main() -> int:
    print(f"Running conformance vectors from {VECTORS_DIR}...")
    print(f"Keys file: {KEYS_FILE} ({'present' if KEYS_FILE.exists() else 'MISSING'})")
    print()
    passed, failed = run_all_vectors()
    total = passed + failed
    print()
    print(f"  {passed}/{total} vectors pass, {failed} failed.")
    return 0 if failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: verifier/typescript/test_vectors.ts
// verifier/typescript/test_vectors.ts
//
// Conformance test runner for the TypeScript reference verifier.
// Loads every vector in vectors/v1/ and asserts that verify() produces
// the same verdict as the Python reference (4/4 pass).
//
// Usage:
//   cd verifier/typescript
//   npm install tweetnacl
//   tsx test_vectors.ts   # or: npx tsx test_vectors.ts
//
// Exit code: 0 on full pass, 1 on any failure. Each failure prints
// the verdict + reason to stderr.
//
// Cross-language conformance: a Go, Rust, or Elixir verifier claiming
// compliance should run the same loop and assert all 4 vectors pass
// with the same verdicts. The vectors are the spec.

import { readFileSync, readdirSync } from "node:fs";
import { join, dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { verify, publicKeyFromHexOrBytes } from "./verify.ts";

const __dirname = dirname(fileURLToPath(import.meta.url));
const VECTORS_DIR = resolve(__dirname, "..", "..", "vectors", "v1");

interface KeysFile {
  [vectorName: string]: {
    public_key_hex: string;
    algorithm: string;
  };
}

function loadKeys(): KeysFile {
  try {
    return JSON.parse(readFileSync(join(VECTORS_DIR, "keys.json"), "utf-8"));
  } catch (e) {
    console.error(`Failed to load ${VECTORS_DIR}/keys.json:`, e);
    return {};
  }
}

function iterVectors(): string[] {
  return readdirSync(VECTORS_DIR)
    .filter((f) => /^\d{2}_.*\.json$/.test(f))
    .sort();
}

function checkOne(vectorName: string, file: string, keys: KeysFile): { ok: boolean; detail: string } {
  const receipt = JSON.parse(file);
  const name = receipt.name;
  if (receipt.version !== 1) {
    return { ok: false, detail: `${name}: expected version=1, got ${receipt.version}` };
  }
  const sigs = receipt.signatures ?? {};
  const sig = sigs.tenant_ed25519;
  if (typeof sig !== "string" || !sig.startsWith("ed25519:")) {
    return { ok: false, detail: `${name}: signatures.tenant_ed25519 is not an ed25519:hex value` };
  }
  if (sigs.cf_attestation != null) {
    return { ok: false, detail: `${name}: cf_attestation must be null in v1` };
  }
  const keyEntry = keys[name];
  if (!keyEntry) {
    return { ok: false, detail: `${name}: missing entry in keys.json` };
  }
  const pub = publicKeyFromHexOrBytes(keyEntry.public_key_hex);
  const verdict = verify(receipt, pub);
  if (!verdict.valid) {
    return { ok: false, detail: `${name}: verify() returned valid=false, reason=${verdict.reason}` };
  }
  return { ok: true, detail: verdict.reason };
}

function main(): number {
  const keys = loadKeys();
  console.log(`Running conformance vectors from ${VECTORS_DIR}`);
  console.log(`Keys file: ${join(VECTORS_DIR, "keys.json")} (${Object.keys(keys).length} entries)`);
  console.log();
  let passed = 0;
  let failed = 0;
  for (const fname of iterVectors()) {
    const fpath = join(VECTORS_DIR, fname);
    const file = readFileSync(fpath, "utf-8");
    const receipt = JSON.parse(file);
    const r = checkOne(receipt.name, file, keys);
    if (r.ok) {
      passed++;
      console.log(`  ✓ ${fname}: ${r.detail}`);
    } else {
      failed++;
      console.error(`  ✗ ${fname}: ${r.detail}`);
    }
  }
  console.log();
  const total = passed + failed;
  console.log(`  ${passed}/${total} vectors pass, ${failed} failed.`);
  return failed === 0 ? 0 : 1;
}

// Export for programmatic use
export { checkOne, loadKeys, iterVectors, main };

if (import.meta.url === `file://${process.argv[1]}`) {
  process.exit(main());
}
